main.go

```diff
@@ -0,0 +1,273 @@
+		_, _ = fmt.Fprintf(hasher, timestamp.Format(time.RFC3339Nano))
+
+		if computed := hasher.Sum32(); chk != computed {
+			return fmt.Errorf("integrity check failed, db.chk=%d, computed hash = %d", chk, computed)
+		}
+
+		m := bigtable.NewMutation()
+		m.Set(cf, col, bigtable.Time(timestamp), value)
+
+		c++
+		keys = append(keys, key)
+		muts = append(muts, m)
+
+		// nothing magic about 100, tune as necessary - we just need some limit so that grpc payloads don't get too huge
+		if len(muts) == 100 {
+			if errs, err := tbl.ApplyBulk(ctx, keys, muts); err != nil {
+				return fmt.Errorf("failed to write to bigtable: %s", err)
+			} else if errs = filterErrors(errs); len(errs) > 0 {
+				return fmt.Errorf("failed to write to bigtable, %d errors, first: %s", len(errs), errs[0])
+			}
+
+			keys = nil
+			muts = nil
+		}
+	}
+
+	if len(muts) > 0 {
+		if errs, err := tbl.ApplyBulk(ctx, keys, muts); err != nil {
+			return fmt.Errorf("failed to write to bigtable: %s", err)
+		} else if errs = filterErrors(errs); len(errs) > 0 {
+			return fmt.Errorf("failed to write to bigtable, %d errors, first: %s", len(errs), errs[0])
+		}
+	}
+
+	log.Printf("restored %d rows for %s", c, table)
+	return nil
+}
+
+func saveAll(ctx context.Context, db *sql.DB, adminClient *bigtable.AdminClient, btClient *bigtable.Client) error {
+	tables, err := adminClient.Tables(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to list BT tables: %s", err)
+	}
+
+	sort.Strings(tables)
+
+	for _, table := range tables {
+		if err := saveTable(ctx, table, db, btClient); err != nil {
+			return fmt.Errorf("failed to save table %q: %s", table, err)
+		}
+	}
+
+	return nil
+}
+
+func saveTable(ctx context.Context, table string, db *sql.DB, btClient *bigtable.Client) error {
+	log.Printf("saving %q table...", table)
+	stmt, err := db.Prepare(`DROP TABLE IF EXISTS "` + table + `"`)
+	if err != nil {
+		return fmt.Errorf("failed to prepare DROP TABLE: %s", err)
+	}
+	if _, err := stmt.Exec(); err != nil {
+		return fmt.Errorf("failed to execute DROP TABLE: %s", err)
+	}
+
+	stmt, err = db.Prepare(`CREATE TABLE IF NOT EXISTS "` + table + `" (key TEXT, column_family TEXT, column TEXT, value BLOB, timestamp INTEGER, chk INTEGER)`)
+	if err != nil {
+		return fmt.Errorf("failed to prepare CREATE TABLE: %s", err)
+	}
+	if _, err := stmt.Exec(); err != nil {
+		return fmt.Errorf("failed to execute CREATE TABLE: %s", err)
+	}
+
+	insertStmt, err := db.Prepare(`INSERT INTO "` + table + `" VALUES (?, ?, ?, ?, ?, ?)`)
+	if err != nil {
+		return fmt.Errorf("failed to prepare INSERT: %s", err)
+	}
+
+	var c int
+	var rowErr error
+	tbl := btClient.Open(table)
+	err = tbl.ReadRows(ctx, bigtable.InfiniteRange(""), func(row bigtable.Row) bool {
+		for cf, items := range row {
+			for _, item := range items {
+				// tricky!  the item.Column that we get back is prefixed with the column family
+				col := strings.TrimPrefix(item.Column, cf+":")
+
+				hasher := fnv.New32a()
```

Comment:
What's up with the integrity check?  Not that it's bad or anything, was just curious what made you want to go this extra mile.

Reply:
Mostly just paranoia.  Since the tool has no knowledge of the shape of your data (everything is just raw bytes as far as its concerned), any kind of error would be very hard to detect.